xla/utils.py

```diff
@@ -11,3 +18,95 @@ def rank_print(fabric: L.Fabric, message: object, *, flush: bool = True, **kwarg
         message = rank_prefixed_message(message, fabric.global_rank)
         # TPU VM will only print when the script finishes if `flush=False`
         print(message, flush=flush, **kwargs)
+
+
+def materialize_parameters(module: torch.nn.Module, device: torch.device)-> None:
+    for module_name, module in module.named_modules():
+        if any(
+            param.is_meta for param in itertools.chain(module.parameters(recurse=False), module.buffers(recurse=False))
+        ):
+            module.to_empty(device=device, recurse=False)
+            module.reset_parameters()
+
+
+def sequential_load_and_fsdp_wrap(
+    fabric: L.Fabric, get_model: Callable[[], GPT], checkpoint_path: Path
+) -> torch.nn.Module:
+    assert fabric._launched
+    # similar logic could be implemented for regular FSDP, but this implementation is specific to XLAFSDP
+    assert isinstance(fabric.strategy, XLAFSDPStrategy)
+
+    with fabric.init_module(empty_init=False), torch.device("meta"):
+        model = get_model()
+
+    # TODO: this could be made faster by broadcasting in separate process groups for each host
+    if fabric.local_rank == 0:
+        # load the full checkpoint on a single rank to limit the system memory usage
+        state_dict = torch.load(checkpoint_path, map_location="cpu", mmap=False)  # mmap=True hangs
```

Comment:
While previously discussed privately, an alternative to `fabric.broadcast` would be to use `torch.dist.broadcast` with `gloo` backend, which is supported on TPU v4+, by using this, you can make minor adjustments to the code to get `mmap=True` to work, as in https://github.com/gkroiz/lit-gpt/blob/benchmarking_clean/finetune/adapter.py. 

We can look more into how to fix the hanging when `mmap=True` and fix in a separate PR.